File: examples/04-game-room/cpp/game-room.cpp
/**
 * Game Room Actor Pattern - 멀티플레이어 게임 룸 예제 in C++
 * CAF 스타일로 구현
 *
 * 컴파일: g++ -std=c++17 game-room.cpp -o game-room -pthread
 */

#include <iostream>
#include <string>
#include <map>
#include <memory>
#include <functional>
#include <thread>
#include <chrono>
#include <mutex>
#include <any>

// ============================================
// 타입 정의
// ============================================

enum class RoomStatus {
    Waiting,
    Starting,
    Playing,
    Finished
};

std::string statusToString(RoomStatus status) {
    switch (status) {
        case RoomStatus::Waiting: return "waiting";
        case RoomStatus::Starting: return "starting";
        case RoomStatus::Playing: return "playing";
        case RoomStatus::Finished: return "finished";
        default: return "unknown";
    }
}

struct PlayerInfo {
    std::string playerId;
    std::string name;
    bool isReady = false;
    int score = 0;
};

// ============================================
// 메시지 타입
// ============================================

struct JoinRoom {
    std::string playerId;
    std::string playerName;
};

struct LeaveRoom {
    std::string playerId;
};

struct SetReady {
    std::string playerId;
    bool ready;
};

struct GameAction {
    std::string playerId;
    std::string actionType;
    int points;
};

struct StartGame {};

struct EndGame {
    std::string winnerId;
};

// Player 메시지
struct PlayerJoined {
    PlayerInfo player;
};

struct PlayerLeft {
    std::string playerId;
};

struct PlayerReadyChanged {
    std::string playerId;
    bool ready;
};

struct GameStarting {
    int countdown;
};

struct GameStarted {};

struct GameActionBroadcast {
    std::string playerId;
    std::string actionType;
    int points;
};

struct GameEnded {
    std::string winnerId;
    std::map<std::string, int> scores;
};

// ============================================
// Player Actor
// ============================================

class PlayerActor {
public:
    PlayerActor(const std::string& playerId, const std::string& name)
        : playerId_(playerId), name_(name) {}

    void receive(const std::any& msg) {
        if (msg.type() == typeid(PlayerJoined)) {
            auto m = std::any_cast<PlayerJoined>(msg);
            std::cout << "[Player " << name_ << "] " << m.player.name << " joined the room\n";
        }
        else if (msg.type() == typeid(PlayerLeft)) {
            auto m = std::any_cast<PlayerLeft>(msg);
            std::cout << "[Player " << name_ << "] Player " << m.playerId << " left\n";
        }
        else if (msg.type() == typeid(PlayerReadyChanged)) {
            auto m = std::any_cast<PlayerReadyChanged>(msg);
            std::cout << "[Player " << name_ << "] Player " << m.playerId
                      << " is " << (m.ready ? "ready" : "not ready") << "\n";
        }
        else if (msg.type() == typeid(GameStarting)) {
            auto m = std::any_cast<GameStarting>(msg);
            std::cout << "[Player " << name_ << "] Game starting in " << m.countdown << "...\n";
        }
        else if (msg.type() == typeid(GameStarted)) {
            std::cout << "[Player " << name_ << "] Game started!\n";
        }
        else if (msg.type() == typeid(GameActionBroadcast)) {
            auto m = std::any_cast<GameActionBroadcast>(msg);
            std::cout << "[Player " << name_ << "] Received action from " << m.playerId << "\n";
        }
        else if (msg.type() == typeid(GameEnded)) {
            auto m = std::any_cast<GameEnded>(msg);
            std::string winner = m.winnerId.empty() ? "none" : m.winnerId;
            std::cout << "[Player " << name_ << "] Game ended! Winner: " << winner << "\n";
        }
    }

private:
    std::string playerId_;
    std::string name_;
};

// ============================================
// Room Actor
// ============================================

class RoomActor {
public:
    explicit RoomActor(const std::string& roomId)
        : roomId_(roomId), status_(RoomStatus::Waiting),
          maxPlayers_(4), minPlayers_(2) {}

    void receive(const std::any& msg) {
        std::lock_guard<std::mutex> lock(mutex_);

        if (msg.type() == typeid(JoinRoom)) {
            handleJoinRoom(std::any_cast<JoinRoom>(msg));
        }
        else if (msg.type() == typeid(LeaveRoom)) {
            handleLeaveRoom(std::any_cast<LeaveRoom>(msg));
        }
        else if (msg.type() == typeid(SetReady)) {
            handleSetReady(std::any_cast<SetReady>(msg));
        }
        else if (msg.type() == typeid(GameAction)) {
            handleGameAction(std::any_cast<GameAction>(msg));
        }
        else if (msg.type() == typeid(StartGame)) {
            handleStartGame();
        }
        else if (msg.type() == typeid(EndGame)) {
            handleEndGame(std::any_cast<EndGame>(msg));
        }
    }

private:
    void handleJoinRoom(const JoinRoom& m) {
        if (players_.size() >= static_cast<size_t>(maxPlayers_)) {
            std::cout << "[Room " << roomId_ << "] Room is full, rejecting " << m.playerName << "\n";
            return;
        }

        if (status_ != RoomStatus::Waiting) {
            std::cout << "[Room " << roomId_ << "] Game in progress, rejecting " << m.playerName << "\n";
            return;
        }

        PlayerInfo player{m.playerId, m.playerName, false, 0};
        players_[m.playerId] = player;

        auto playerActor = std::make_shared<PlayerActor>(m.playerId, m.playerName);
        playerActors_[m.playerId] = playerActor;

        std::cout << "[Room " << roomId_ << "] " << m.playerName
                  << " joined (" << players_.size() << "/" << maxPlayers_ << ")\n";

        broadcast(m.playerId, PlayerJoined{player});
    }

    void handleLeaveRoom(const LeaveRoom& m) {
        auto it = players_.find(m.playerId);
        if (it == players_.end()) return;

        std::string playerName = it->second.name;
        players_.erase(it);
        playerActors_.erase(m.playerId);

        std::cout << "[Room " << roomId_ << "] " << playerName
                  << " left (" << players_.size() << "/" << maxPlayers_ << ")\n";

        broadcast("", PlayerLeft{m.playerId});

        if (status_ == RoomStatus::Playing && players_.size() < static_cast<size_t>(minPlayers_)) {
            status_ = RoomStatus::Finished;
            std::cout << "[Room " << roomId_ << "] Not enough players, game ended\n";
            broadcast("", GameEnded{"", getScores()});
        }

        if (players_.empty()) {
            status_ = RoomStatus::Waiting;
            std::cout << "[Room " << roomId_ << "] Room is empty, resetting\n";
        }
    }

    void handleSetReady(const SetReady& m) {
        auto it = players_.find(m.playerId);
        if (it == players_.end() || status_ != RoomStatus::Waiting) return;

        it->second.isReady = m.ready;
        std::cout << "[Room " << roomId_ << "] " << it->second.name
                  << " is " << (m.ready ? "ready" : "not ready") << "\n";

        broadcast("", PlayerReadyChanged{m.playerId, m.ready});
        checkStartCondition();
    }

    void handleGameAction(const GameAction& m) {
        if (status_ != RoomStatus::Playing) return;

        auto it = players_.find(m.playerId);
        if (it == players_.end()) return;

        std::cout << "[Room " << roomId_ << "] " << it->second.name
                  << " action: " << m.actionType << "\n";

        if (m.actionType == "SCORE") {
            it->second.score += m.points;
            std::cout << "[Room " << roomId_ << "] " << it->second.name
                      << " score: " << it->second.score << "\n";
        }

        broadcast("", GameActionBroadcast{m.playerId, m.actionType, m.points});
    }

    void handleStartGame() {
        if (status_ != RoomStatus::Waiting) return;

        status_ = RoomStatus::Starting;
        std::cout << "[Room " << roomId_ << "] Game starting...\n";

        std::thread([this]() {
            for (int countdown = 3; countdown > 0; countdown--) {
                {
                    std::lock_guard<std::mutex> lock(mutex_);
                    broadcast("", GameStarting{countdown});
                }
                std::cout << "[Room " << roomId_ << "] Starting in " << countdown << "...\n";
                std::this_thread::sleep_for(std::chrono::seconds(1));
            }

            {
                std::lock_guard<std::mutex> lock(mutex_);
                status_ = RoomStatus::Playing;
                std::cout << "[Room " << roomId_ << "] Game started!\n";
                broadcast("", GameStarted{});
            }
        }).detach();
    }

    void handleEndGame(const EndGame& m) {
        if (status_ != RoomStatus::Playing) return;

        status_ = RoomStatus::Finished;
        std::string winner = m.winnerId.empty() ? "none" : m.winnerId;
        std::cout << "[Room " << roomId_ << "] Game ended! Winner: " << winner << "\n";

        broadcast("", GameEnded{m.winnerId, getScores()});

        std::thread([this]() {
            std::this_thread::sleep_for(std::chrono::seconds(5));
            std::lock_guard<std::mutex> lock(mutex_);
            status_ = RoomStatus::Waiting;
            for (auto& [_, player] : players_) {
                player.isReady = false;
                player.score = 0;
            }
            std::cout << "[Room " << roomId_ << "] Room reset to waiting\n";
        }).detach();
    }

    template<typename T>
    void broadcast(const std::string& excludeId, const T& msg) {
        for (const auto& [id, actor] : playerActors_) {
            if (id != excludeId) {
                actor->receive(msg);
            }
        }
    }

    void checkStartCondition() {
        if (players_.size() >= static_cast<size_t>(minPlayers_)) {
            bool allReady = true;
            for (const auto& [_, player] : players_) {
                if (!player.isReady) {
                    allReady = false;
                    break;
                }
            }
            if (allReady) {
                std::cout << "[Room " << roomId_ << "] All players ready!\n";
                handleStartGame();
            }
        }
    }

    std::map<std::string, int> getScores() const {
        std::map<std::string, int> scores;
        for (const auto& [id, player] : players_) {
            scores[id] = player.score;
        }
        return scores;
    }

    std::string roomId_;
    RoomStatus status_;
    std::map<std::string, PlayerInfo> players_;
    std::map<std::string, std::shared_ptr<PlayerActor>> playerActors_;
    int maxPlayers_;
    int minPlayers_;
    std::mutex mutex_;
};

// ============================================
// 메인 함수
// ============================================

int main() {
    std::cout << "=== Game Room Actor Pattern Demo (C++) ===\n\n";

    RoomActor room("001");

    std::cout << "--- 1. Players Joining ---\n\n";

    room.receive(JoinRoom{"p1", "Alice"});
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    room.receive(JoinRoom{"p2", "Bob"});
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    room.receive(JoinRoom{"p3", "Charlie"});
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    std::cout << "\n--- 2. Players Getting Ready ---\n\n";

    room.receive(SetReady{"p1", true});
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    room.receive(SetReady{"p2", true});
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    // 마지막 플레이어가 준비되면 게임 시작
    room.receive(SetReady{"p3", true});

    // 게임 시작 대기
    std::this_thread::sleep_for(std::chrono::seconds(5));

    std::cout << "\n--- 3. Game Actions ---\n\n";

    room.receive(GameAction{"p1", "SCORE", 10});
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    room.receive(GameAction{"p2", "SCORE", 15});
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    room.receive(GameAction{"p1", "SCORE", 20});
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    std::cout << "\n--- 4. Game End ---\n\n";

    room.receive(EndGame{"p1"});
    std::this_thread::sleep_for(std::chrono::seconds(1));

    std::cout << "\n--- 5. Player Leaving ---\n\n";

    room.receive(LeaveRoom{"p3"});
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    std::cout << "\n=== Demo Complete ===\n";

    return 0;
}

File: examples/04-game-room/go/game-room.go
// Game Room Actor Pattern - 멀티플레이어 게임 룸 예제 in Go
// Proto.Actor 스타일로 구현

package main

import (
	"fmt"
	"sync"
	"time"
)

// ============================================
// 타입 정의
// ============================================

type RoomStatus string

const (
	StatusWaiting  RoomStatus = "waiting"
	StatusStarting RoomStatus = "starting"
	StatusPlaying  RoomStatus = "playing"
	StatusFinished RoomStatus = "finished"
)

type PlayerInfo struct {
	PlayerId string
	Name     string
	IsReady  bool
	Score    int
}

type RoomState struct {
	RoomId     string
	Status     RoomStatus
	Players    map[string]*PlayerInfo
	MaxPlayers int
	MinPlayers int
}

// ============================================
// 메시지 타입
// ============================================

// Room 메시지
type JoinRoom struct {
	PlayerId   string
	PlayerName string
}

type LeaveRoom struct {
	PlayerId string
}

type SetReady struct {
	PlayerId string
	Ready    bool
}

type GameAction struct {
	PlayerId   string
	ActionType string
	Points     int
}

type StartGame struct{}
type EndGame struct {
	WinnerId string
}
type GetState struct {
	Reply chan *RoomState
}

// Player 메시지
type PlayerJoined struct {
	Player *PlayerInfo
}

type PlayerLeft struct {
	PlayerId string
}

type PlayerReadyChanged struct {
	PlayerId string
	Ready    bool
}

type GameStarting struct {
	Countdown int
}

type GameStarted struct{}

type GameActionBroadcast struct {
	PlayerId   string
	ActionType string
	Points     int
}

type GameEnded struct {
	WinnerId string
	Scores   map[string]int
}

// ============================================
// Room Actor
// ============================================

type RoomActor struct {
	roomId       string
	status       RoomStatus
	players      map[string]*PlayerInfo
	playerActors map[string]*PlayerActor
	maxPlayers   int
	minPlayers   int
	mu           sync.RWMutex
}

func NewRoomActor(roomId string) *RoomActor {
	return &RoomActor{
		roomId:       roomId,
		status:       StatusWaiting,
		players:      make(map[string]*PlayerInfo),
		playerActors: make(map[string]*PlayerActor),
		maxPlayers:   4,
		minPlayers:   2,
	}
}

func (r *RoomActor) Receive(msg interface{}) {
	r.mu.Lock()
	defer r.mu.Unlock()

	switch m := msg.(type) {
	case JoinRoom:
		r.handleJoinRoom(m)
	case LeaveRoom:
		r.handleLeaveRoom(m)
	case SetReady:
		r.handleSetReady(m)
	case GameAction:
		r.handleGameAction(m)
	case StartGame:
		r.handleStartGame()
	case EndGame:
		r.handleEndGame(m)
	case GetState:
		r.handleGetState(m)
	}
}

func (r *RoomActor) handleJoinRoom(m JoinRoom) {
	if len(r.players) >= r.maxPlayers {
		fmt.Printf("[Room %s] Room is full, rejecting %s\n", r.roomId, m.PlayerName)
		return
	}

	if r.status != StatusWaiting {
		fmt.Printf("[Room %s] Game in progress, rejecting %s\n", r.roomId, m.PlayerName)
		return
	}

	player := &PlayerInfo{
		PlayerId: m.PlayerId,
		Name:     m.PlayerName,
		IsReady:  false,
		Score:    0,
	}
	r.players[m.PlayerId] = player

	playerActor := NewPlayerActor(m.PlayerId, m.PlayerName)
	r.playerActors[m.PlayerId] = playerActor

	fmt.Printf("[Room %s] %s joined (%d/%d)\n", r.roomId, m.PlayerName, len(r.players), r.maxPlayers)

	r.broadcast(m.PlayerId, PlayerJoined{Player: player})
}

func (r *RoomActor) handleLeaveRoom(m LeaveRoom) {
	player, exists := r.players[m.PlayerId]
	if !exists {
		return
	}

	delete(r.players, m.PlayerId)
	delete(r.playerActors, m.PlayerId)

	fmt.Printf("[Room %s] %s left (%d/%d)\n", r.roomId, player.Name, len(r.players), r.maxPlayers)

	r.broadcast("", PlayerLeft{PlayerId: m.PlayerId})

	if r.status == StatusPlaying && len(r.players) < r.minPlayers {
		r.status = StatusFinished
		fmt.Printf("[Room %s] Not enough players, game ended\n", r.roomId)
		r.broadcast("", GameEnded{Scores: r.getScores()})
	}

	if len(r.players) == 0 {
		r.status = StatusWaiting
		fmt.Printf("[Room %s] Room is empty, resetting\n", r.roomId)
	}
}

func (r *RoomActor) handleSetReady(m SetReady) {
	player, exists := r.players[m.PlayerId]
	if !exists || r.status != StatusWaiting {
		return
	}

	player.IsReady = m.Ready
	readyStr := "not ready"
	if m.Ready {
		readyStr = "ready"
	}
	fmt.Printf("[Room %s] %s is %s\n", r.roomId, player.Name, readyStr)

	r.broadcast("", PlayerReadyChanged{PlayerId: m.PlayerId, Ready: m.Ready})
	r.checkStartCondition()
}

func (r *RoomActor) handleGameAction(m GameAction) {
	if r.status != StatusPlaying {
		return
	}

	player, exists := r.players[m.PlayerId]
	if !exists {
		return
	}

	fmt.Printf("[Room %s] %s action: %s\n", r.roomId, player.Name, m.ActionType)

	if m.ActionType == "SCORE" {
		player.Score += m.Points
		fmt.Printf("[Room %s] %s score: %d\n", r.roomId, player.Name, player.Score)
	}

	r.broadcast("", GameActionBroadcast{
		PlayerId:   m.PlayerId,
		ActionType: m.ActionType,
		Points:     m.Points,
	})
}

func (r *RoomActor) handleStartGame() {
	if r.status != StatusWaiting {
		return
	}

	r.status = StatusStarting
	fmt.Printf("[Room %s] Game starting...\n", r.roomId)

	go func() {
		for countdown := 3; countdown > 0; countdown-- {
			r.mu.RLock()
			r.broadcast("", GameStarting{Countdown: countdown})
			r.mu.RUnlock()
			fmt.Printf("[Room %s] Starting in %d...\n", r.roomId, countdown)
			time.Sleep(time.Second)
		}

		r.mu.Lock()
		r.status = StatusPlaying
		fmt.Printf("[Room %s] Game started!\n", r.roomId)
		r.broadcast("", GameStarted{})
		r.mu.Unlock()
	}()
}

func (r *RoomActor) handleEndGame(m EndGame) {
	if r.status != StatusPlaying {
		return
	}

	r.status = StatusFinished
	winner := m.WinnerId
	if winner == "" {
		winner = "none"
	}
	fmt.Printf("[Room %s] Game ended! Winner: %s\n", r.roomId, winner)

	r.broadcast("", GameEnded{WinnerId: m.WinnerId, Scores: r.getScores()})

	go func() {
		time.Sleep(5 * time.Second)
		r.mu.Lock()
		r.status = StatusWaiting
		for _, player := range r.players {
			player.IsReady = false
			player.Score = 0
		}
		fmt.Printf("[Room %s] Room reset to waiting\n", r.roomId)
		r.mu.Unlock()
	}()
}

func (r *RoomActor) handleGetState(m GetState) {
	state := &RoomState{
		RoomId:     r.roomId,
		Status:     r.status,
		Players:    r.players,
		MaxPlayers: r.maxPlayers,
		MinPlayers: r.minPlayers,
	}
	m.Reply <- state
}

func (r *RoomActor) broadcast(excludeId string, msg interface{}) {
	for id, actor := range r.playerActors {
		if id != excludeId {
			actor.Receive(msg)
		}
	}
}

func (r *RoomActor) checkStartCondition() {
	if len(r.players) >= r.minPlayers {
		allReady := true
		for _, p := range r.players {
			if !p.IsReady {
				allReady = false
				break
			}
		}
		if allReady {
			fmt.Printf("[Room %s] All players ready!\n", r.roomId)
			go r.handleStartGame()
		}
	}
}

func (r *RoomActor) getScores() map[string]int {
	scores := make(map[string]int)
	for id, player := range r.players {
		scores[id] = player.Score
	}
	return scores
}

// ============================================
// Player Actor
// ============================================

type PlayerActor struct {
	playerId string
	name     string
}

func NewPlayerActor(playerId, name string) *PlayerActor {
	return &PlayerActor{
		playerId: playerId,
		name:     name,
	}
}

func (p *PlayerActor) Receive(msg interface{}) {
	switch m := msg.(type) {
	case PlayerJoined:
		fmt.Printf("[Player %s] %s joined the room\n", p.name, m.Player.Name)
	case PlayerLeft:
		fmt.Printf("[Player %s] Player %s left\n", p.name, m.PlayerId)
	case PlayerReadyChanged:
		readyStr := "not ready"
		if m.Ready {
			readyStr = "ready"
		}
		fmt.Printf("[Player %s] Player %s is %s\n", p.name, m.PlayerId, readyStr)
	case GameStarting:
		fmt.Printf("[Player %s] Game starting in %d...\n", p.name, m.Countdown)
	case GameStarted:
		fmt.Printf("[Player %s] Game started!\n", p.name)
	case GameActionBroadcast:
		fmt.Printf("[Player %s] Received action from %s\n", p.name, m.PlayerId)
	case GameEnded:
		winner := m.WinnerId
		if winner == "" {
			winner = "none"
		}
		fmt.Printf("[Player %s] Game ended! Winner: %s\n", p.name, winner)
	}
}

// ============================================
// 메인 함수
// ============================================

func main() {
	fmt.Println("=== Game Room Actor Pattern Demo (Go) ===\n")

	room := NewRoomActor("001")

	fmt.Println("--- 1. Players Joining ---\n")

	room.Receive(JoinRoom{PlayerId: "p1", PlayerName: "Alice"})
	time.Sleep(100 * time.Millisecond)

	room.Receive(JoinRoom{PlayerId: "p2", PlayerName: "Bob"})
	time.Sleep(100 * time.Millisecond)

	room.Receive(JoinRoom{PlayerId: "p3", PlayerName: "Charlie"})
	time.Sleep(100 * time.Millisecond)

	fmt.Println("\n--- 2. Players Getting Ready ---\n")

	room.Receive(SetReady{PlayerId: "p1", Ready: true})
	time.Sleep(100 * time.Millisecond)

	room.Receive(SetReady{PlayerId: "p2", Ready: true})
	time.Sleep(100 * time.Millisecond)

	// 마지막 플레이어가 준비되면 게임 시작
	room.Receive(SetReady{PlayerId: "p3", Ready: true})

	// 게임 시작 대기
	time.Sleep(5 * time.Second)

	fmt.Println("\n--- 3. Game Actions ---\n")

	room.Receive(GameAction{PlayerId: "p1", ActionType: "SCORE", Points: 10})
	time.Sleep(100 * time.Millisecond)

	room.Receive(GameAction{PlayerId: "p2", ActionType: "SCORE", Points: 15})
	time.Sleep(100 * time.Millisecond)

	room.Receive(GameAction{PlayerId: "p1", ActionType: "SCORE", Points: 20})
	time.Sleep(100 * time.Millisecond)

	fmt.Println("\n--- 4. Game End ---\n")

	room.Receive(EndGame{WinnerId: "p1"})
	time.Sleep(time.Second)

	fmt.Println("\n--- 5. Player Leaving ---\n")

	room.Receive(LeaveRoom{PlayerId: "p3"})
	time.Sleep(100 * time.Millisecond)

	fmt.Println("\n=== Demo Complete ===")
}
